libr/core/rvc.c

```diff
@@ -66,491 +32,306 @@ static inline char *sha256_file(const char *fname) {
-	nc->message = r_str_new (message);
-	if (!nc->message) {
-		free (nc->author);
-		free (nc);
+	commit_hash = find_sha256 ((unsigned char *)
+			content, r_str_len_utf8 (content));
+	if (!commit_hash) {
+		free (content);
 		return false;
 	}
-	nc->timestamp = time (NULL);
-	nc->prev = repo->current_branch->head;
-	nc->blobs = blobs;
-	if (nc->prev) {
-		nc->prev->next = nc;
-	}
-	if (!write_commit (repo, repo->current_branch, nc)) {
-		free (nc->author);
-		free (nc->message);
-		free (nc);
-		eprintf ("Failed To Create Commit File\n");
+	commit_path = r_str_newf ("%s" R_SYS_DIR ".rvc" R_SYS_DIR "commits"
+			R_SYS_DIR "%s", rp, commit_hash);
+	if (!commit_path) {
+		free (content);
+		free (commit_hash);
 		return false;
 	}
-	repo->current_branch->head = nc;
-	return true;
-}
-
-R_API bool r_vc_branch(Rvc *repo, const char *name) {
-	char *bpath, *ppath;
-	RvcBranch *nb;
-	if (!is_branch_name (name)) {
-		eprintf ("%s Is Not A Vaild Branch Name\n", name);
+	commitf = fopen (commit_path, "w+");
+	free (commit_path);
+	if (!commitf) {
+		free (content);
+		free (commit_hash);
 		return false;
 	}
-	nb = R_NEW0 (RvcBranch);
-	if (!nb) {
-		eprintf ("Failed To Allocate Branch Struct\n");
+	if (fprintf (commitf, "%s", content) != r_str_len_utf8 (content) * sizeof (char)) {
+		free (content);
+		free (commit_hash);
+		fclose (commitf);
 		return false;
 	}
-	nb->head = NULL;
-	nb->name = r_str_new (name);
-	if (!nb->name) {
-		eprintf ("Failed To Allocate Branch Name\n");
-		free (nb);
+	fclose (commitf);
+	free (content);
+	return commit_hash;
+
+
+}
+
+R_API bool r_vc_commit(const char *rp, const char *message, const char *auth, RList *files) {
+	char *commit_hash;
+	RList *blobs = blobs_add (files, rp);
+	if (!blobs) {
 		return false;
 	}
-	r_list_append (repo->branches, nb);
-	bpath = branch_mkdir (repo, nb);
-	if (!bpath) {
-		eprintf ("Failed To Create Branch Directory\n");
-		free (nb->name);
-		free (nb);
-		r_list_pop (repo->branches);
+	commit_hash = write_commit (rp, message, auth, blobs);
+	if (!commit_hash) {
+		free_blobs (blobs);
 		return false;
 	}
-	if (repo->current_branch) {
-		nb->head = repo->current_branch->head;
-		ppath = r_str_newf ("%s" R_SYS_DIR "branches" R_SYS_DIR "%s"
-				R_SYS_DIR "commits" R_SYS_DIR,
-				repo->path, repo->current_branch->name);
-		if (!copy_commits (repo, bpath, ppath)) {
-			eprintf ("Failed To Copy Commits From Parent\n");
-			free (nb->name);
-			free (nb);
-			free (bpath);
-			free (ppath);
+	{
+		Sdb *db = sdb_new0 ();
+		char *dbf;
+		char *current_branch;
+		if (!db) {
+			free_blobs (blobs);
+			free (commit_hash);
+			return false;
+		}
+		dbf = r_str_newf ("%s" R_SYS_DIR ".rvc" R_SYS_DIR "branches.sdb",
+				rp);
+		if (!dbf) {
+			free_blobs (blobs);
+			free (commit_hash);
+			return false;
+		}
+		if (sdb_open (db, dbf) < 0) {
+			free_blobs (blobs);
+			free (commit_hash);
 			return false;
 		}
-		free (ppath);
+		free (dbf);
+		current_branch = sdb_get (db, "current_branch", 0);
```

Comment:
you can use `sdb_const_get` and avoid the free later. if you dont modify the value it will save a dupped string in memory